File: src/grok/spawn.ts
import { spawn, type ChildProcess } from "node:child_process";
import { existsSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";

export const GROK_BIN = "grok";
export const GROK_ACP_ARGS = ["--no-auto-update", "--disallowed-tools", "search_tool,Agent,run_terminal_command", "agent", "--always-approve", "stdio"] as const;

function grokDirIfPresent(dir: string): string[] {
  return existsSync(join(dir, "grok")) ? [dir] : [];
}

export function pathWithLocalGrok(env: NodeJS.ProcessEnv): string {
  const current = env.PATH ?? process.env.PATH ?? "";
  const home = env.HOME ?? process.env.HOME ?? homedir();
  const prefix = [
    ...grokDirIfPresent(join(home, ".grok", "bin")),
    ...grokDirIfPresent(join(home, ".local", "bin")),
  ];
  return [...prefix, current].join(":");
}

export function spawnGrok(env: NodeJS.ProcessEnv): ChildProcess {
  const child = spawn(GROK_BIN, [...GROK_ACP_ARGS], {
    stdio: ["pipe", "pipe", "pipe"],
    env: { ...env, PATH: pathWithLocalGrok(env), GROK_SUBAGENTS: "0" },
  });
  child.on("error", (err) => {
    process.stderr.write(`[host] grok spawn: ${err.message}\n`);
  });
  return child;
}

File: python/teams_harness/spawn.py
from __future__ import annotations

import os
import shutil
import sys
from collections.abc import Mapping
from pathlib import Path
from subprocess import PIPE, Popen

GROK_BIN = "grok"
GROK_ACP_ARGS: tuple[str, ...] = (
    "--no-auto-update",
    "--disallowed-tools",
    "search_tool,Agent,run_terminal_command",
    "agent",
    "--always-approve",
    "stdio",
)


def _grok_dir_if_present(directory: Path) -> list[str]:
    return [str(directory)] if (directory / "grok").exists() else []


def path_with_local_grok(env: Mapping[str, str] | None = None) -> str:
    source = env if env is not None else os.environ
    current = source.get("PATH") or os.environ.get("PATH") or ""
    home = source.get("HOME") or os.environ.get("HOME") or str(Path.home())
    prefix = [
        *_grok_dir_if_present(Path(home) / ".grok" / "bin"),
        *_grok_dir_if_present(Path(home) / ".local" / "bin"),
    ]
    return ":".join([*prefix, current]) if prefix else current


def spawn_grok(env: Mapping[str, str] | None = None) -> Popen[bytes]:
    merged = {**os.environ, **(dict(env) if env is not None else {})}
    merged["PATH"] = path_with_local_grok(merged)
    merged["GROK_SUBAGENTS"] = "0"
    try:
        child = Popen(  # noqa: S603 — production spawn is always the real grok CLI
            [GROK_BIN, *GROK_ACP_ARGS],
            stdin=PIPE,
            stdout=PIPE,
            stderr=PIPE,
            env=merged,
        )
    except OSError as err:
        sys.stderr.write(f"[host] grok spawn: {err}\n")
        raise
    return child


def which_grok(env: Mapping[str, str] | None = None) -> str | None:
    path = path_with_local_grok(env)
    return shutil.which(GROK_BIN, path=path)
